File: bindings/rust/src/git.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//

//! Git Server types for the proven-servers ABI.
//!
//! Formally verified Git smart protocol types.
//! Mirrors the Idris2 module `GitABI.Types`.
//!
//! - `Command` -- Git protocol commands.
//! - `PacketType` -- Git protocol packet types.
//! - `RefType` -- Git reference types.
//! - `Capability` -- Git protocol capabilities.
//! - `HookResult` -- Git hook results.
//! - `ServerState` -- Git server states.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// Git Server Constants
// ===========================================================================

/// Standard Git daemon port.
pub const GIT_PORT: u16 = 9418;

// ===========================================================================
// Command (tags 0-2)
// ===========================================================================

/// Git protocol commands.
///
/// Matches `Command` in `GitABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Command {
    /// git-upload-pack (tag 0).
    UploadPack = 0,
    /// git-receive-pack (tag 1).
    ReceivePack = 1,
    /// git-upload-archive (tag 2).
    UploadArchive = 2,
}

impl Command {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::UploadPack),
            1 => Some(Self::ReceivePack),
            2 => Some(Self::UploadArchive),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [Command; 3] = [
        Self::UploadPack, Self::ReceivePack, Self::UploadArchive,
    ];
}

impl fmt::Display for Command {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// PacketType (tags 0-7)
// ===========================================================================

/// Git protocol packet types.
///
/// Matches `PacketType` in `GitABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum PacketType {
    /// Flush (tag 0).
    Flush = 0,
    /// Delimiter (tag 1).
    Delimiter = 1,
    /// ResponseEnd (tag 2).
    ResponseEnd = 2,
    /// Data (tag 3).
    Data = 3,
    /// Error packet (tag 4).
    PktError = 4,
    /// SidebandData (tag 5).
    SidebandData = 5,
    /// SidebandProgress (tag 6).
    SidebandProgress = 6,
    /// SidebandError (tag 7).
    SidebandError = 7,
}

impl PacketType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Flush),
            1 => Some(Self::Delimiter),
            2 => Some(Self::ResponseEnd),
            3 => Some(Self::Data),
            4 => Some(Self::PktError),
            5 => Some(Self::SidebandData),
            6 => Some(Self::SidebandProgress),
            7 => Some(Self::SidebandError),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [PacketType; 8] = [
        Self::Flush, Self::Delimiter, Self::ResponseEnd, Self::Data, Self::PktError, Self::SidebandData, Self::SidebandProgress, Self::SidebandError,
    ];
}

impl fmt::Display for PacketType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// RefType (tags 0-4)
// ===========================================================================

/// Git reference types.
///
/// Matches `RefType` in `GitABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum RefType {
    /// Branch (tag 0).
    Branch = 0,
    /// Tag (tag 1).
    Tag = 1,
    /// Head (tag 2).
    Head = 2,
    /// Remote (tag 3).
    Remote = 3,
    /// Note (tag 4).
    GitNote = 4,
}

impl RefType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Branch),
            1 => Some(Self::Tag),
            2 => Some(Self::Head),
            3 => Some(Self::Remote),
            4 => Some(Self::GitNote),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [RefType; 5] = [
        Self::Branch, Self::Tag, Self::Head, Self::Remote, Self::GitNote,
    ];
}

impl fmt::Display for RefType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Capability (tags 0-8)
// ===========================================================================

/// Git protocol capabilities.
///
/// Matches `Capability` in `GitABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Capability {
    /// MultiAck (tag 0).
    MultiAck = 0,
    /// ThinPack (tag 1).
    ThinPack = 1,
    /// SideBand64k (tag 2).
    SideBand64k = 2,
    /// OFS-delta (tag 3).
    OfsDelta = 3,
    /// Shallow (tag 4).
    Shallow = 4,
    /// DeepenSince (tag 5).
    DeepenSince = 5,
    /// DeepenNot (tag 6).
    DeepenNot = 6,
    /// FilterSpec (tag 7).
    FilterSpec = 7,
    /// ObjectFormat (tag 8).
    ObjectFormat = 8,
}

impl Capability {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::MultiAck),
            1 => Some(Self::ThinPack),
            2 => Some(Self::SideBand64k),
            3 => Some(Self::OfsDelta),
            4 => Some(Self::Shallow),
            5 => Some(Self::DeepenSince),
            6 => Some(Self::DeepenNot),
            7 => Some(Self::FilterSpec),
            8 => Some(Self::ObjectFormat),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [Capability; 9] = [
        Self::MultiAck, Self::ThinPack, Self::SideBand64k, Self::OfsDelta, Self::Shallow, Self::DeepenSince, Self::DeepenNot, Self::FilterSpec, Self::ObjectFormat,
    ];
}

impl fmt::Display for Capability {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// HookResult (tags 0-1)
// ===========================================================================

/// Git hook results.
///
/// Matches `HookResult` in `GitABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum HookResult {
    /// Accept (tag 0).
    Accept = 0,
    /// Reject (tag 1).
    Reject = 1,
}

impl HookResult {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Accept),
            1 => Some(Self::Reject),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [HookResult; 2] = [
        Self::Accept, Self::Reject,
    ];
}

impl fmt::Display for HookResult {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ServerState (tags 0-4)
// ===========================================================================

/// Git server states.
///
/// Matches `ServerState` in `GitABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ServerState {
    /// Idle (tag 0).
    Idle = 0,
    /// Discovery (tag 1).
    Discovery = 1,
    /// Negotiating (tag 2).
    Negotiating = 2,
    /// Transfer (tag 3).
    Transfer = 3,
    /// Shutdown (tag 4).
    Shutdown = 4,
}

impl ServerState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Idle),
            1 => Some(Self::Discovery),
            2 => Some(Self::Negotiating),
            3 => Some(Self::Transfer),
            4 => Some(Self::Shutdown),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ServerState; 5] = [
        Self::Idle, Self::Discovery, Self::Negotiating, Self::Transfer, Self::Shutdown,
    ];
}

impl fmt::Display for ServerState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn command_roundtrip() {
        for v in Command::ALL {
            let tag = v.to_tag();
            let decoded = Command::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Command::from_tag(3).is_none());
    }

    #[test]
    fn packet_type_roundtrip() {
        for v in PacketType::ALL {
            let tag = v.to_tag();
            let decoded = PacketType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(PacketType::from_tag(8).is_none());
    }

    #[test]
    fn ref_type_roundtrip() {
        for v in RefType::ALL {
            let tag = v.to_tag();
            let decoded = RefType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(RefType::from_tag(5).is_none());
    }

    #[test]
    fn capability_roundtrip() {
        for v in Capability::ALL {
            let tag = v.to_tag();
            let decoded = Capability::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Capability::from_tag(9).is_none());
    }

    #[test]
    fn hook_result_roundtrip() {
        for v in HookResult::ALL {
            let tag = v.to_tag();
            let decoded = HookResult::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(HookResult::from_tag(2).is_none());
    }

    #[test]
    fn server_state_roundtrip() {
        for v in ServerState::ALL {
            let tag = v.to_tag();
            let decoded = ServerState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ServerState::from_tag(5).is_none());
    }

    #[test]
    fn constants_match_idris() {
        assert_eq!(GIT_PORT, 9418);
    }

}

File: protocols/proven-git/ffi/zig/src/git.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// git.zig -- Zig FFI implementation of proven-git.
//
// Implements the Git smart transport server state machine with:
//   - 64-slot mutex-protected session pool
//   - Ref advertisement tracking
//   - Pack negotiation lifecycle
//   - Hook execution simulation
//   - Thread-safe via per-pool mutex

const std = @import("std");

pub const Command = enum(u8) { upload_pack = 0, receive_pack = 1, upload_archive = 2 };
pub const PacketType = enum(u8) { flush = 0, delimiter = 1, response_end = 2, data = 3, pkt_error = 4, sideband_data = 5, sideband_progress = 6, sideband_error = 7 };
pub const RefType = enum(u8) { branch = 0, tag = 1, head = 2, remote = 3, note = 4 };
pub const Capability = enum(u8) { multi_ack = 0, thin_pack = 1, side_band_64k = 2, ofs_delta = 3, shallow = 4, deepen_since = 5, deepen_not = 6, filter_spec = 7, object_format = 8 };
pub const HookResult = enum(u8) { accept = 0, reject = 1 };
pub const ServerState = enum(u8) { idle = 0, discovery = 1, negotiating = 2, transfer = 3, shutdown = 4 };

const MAX_SESSIONS: usize = 64;
const MAX_NAME_LEN: usize = 256;
const MAX_REFS: usize = 256;

const Ref = struct { name: [MAX_NAME_LEN]u8, name_len: u32, ref_type: RefType, active: bool };
const empty_ref: Ref = .{ .name = [_]u8{0} ** MAX_NAME_LEN, .name_len = 0, .ref_type = .branch, .active = false };

const Session = struct {
    state: ServerState, path: [MAX_NAME_LEN]u8, path_len: u32, command: Command,
    refs: [MAX_REFS]Ref, ref_count: u32, active: bool,
};
const empty_session: Session = .{
    .state = .idle, .path = [_]u8{0} ** MAX_NAME_LEN, .path_len = 0, .command = .upload_pack,
    .refs = [_]Ref{empty_ref} ** MAX_REFS, .ref_count = 0, .active = false,
};

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

pub export fn git_abi_version() callconv(.c) u32 { return 1; }

pub export fn git_create(path_ptr: [*]const u8, path_len: u32, cmd: u8) callconv(.c) c_int {
    mutex.lock(); defer mutex.unlock();
    if (path_len == 0 or path_len > MAX_NAME_LEN) return -1;
    if (cmd > 2) return -1;
    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            @memcpy(s.path[0..path_len], path_ptr[0..path_len]);
            s.path_len = path_len; s.command = @enumFromInt(cmd);
            s.state = .discovery; s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

pub export fn git_destroy(slot: c_int) callconv(.c) void {
    mutex.lock(); defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

pub export fn git_state(slot: c_int) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

pub export fn git_advertise_ref(slot: c_int, ref_type: u8, name_ptr: [*]const u8, name_len: u32) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .discovery) return 1;
    if (ref_type > 4) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;
    for (&sessions[idx].refs) |*r| {
        if (!r.active) {
            @memcpy(r.name[0..name_len], name_ptr[0..name_len]);
            r.name_len = name_len; r.ref_type = @enumFromInt(ref_type); r.active = true;
            sessions[idx].ref_count += 1;
            return 0;
        }
    }
    return 1;
}

pub export fn git_ref_count(slot: c_int) callconv(.c) u32 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].ref_count;
}

pub export fn git_begin_negotiation(slot: c_int) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .discovery) return 1;
    sessions[idx].state = .negotiating; return 0;
}

pub export fn git_finish_negotiation(slot: c_int) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .negotiating) return 1;
    sessions[idx].state = .transfer; return 0;
}

pub export fn git_begin_transfer(slot: c_int) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .negotiating) return 1;
    sessions[idx].state = .transfer; return 0;
}

pub export fn git_finish_transfer(slot: c_int) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .transfer) return 1;
    sessions[idx].state = .idle; return 0;
}

pub export fn git_run_hook(slot: c_int, hook_type: u8) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    _ = hook_type;
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .transfer and sessions[idx].state != .negotiating) return 1;
    return 0; // accept
}

pub export fn git_shutdown(slot: c_int) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state == .discovery or state == .negotiating or state == .transfer) {
        sessions[idx].state = .shutdown; return 0;
    }
    return 1;
}

pub export fn git_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock(); defer mutex.unlock();
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .shutdown) return 1;
    sessions[idx].state = .idle; sessions[idx].refs = [_]Ref{empty_ref} ** MAX_REFS;
    sessions[idx].ref_count = 0; return 0;
}

pub export fn git_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1;
    if (from == 1 and to == 2) return 1;
    if (from == 2 and to == 3) return 1;
    if (from == 3 and to == 0) return 1;
    if (from == 1 and to == 4) return 1;
    if (from == 2 and to == 4) return 1;
    if (from == 3 and to == 4) return 1;
    if (from == 4 and to == 0) return 1;
    return 0;
}
